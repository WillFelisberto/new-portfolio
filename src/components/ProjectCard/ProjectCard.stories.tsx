import { Meta, StoryFn } from '@storybook/react';
import ProjectCard from '.';

export default {
  title: 'Components/ProjectCard',
  component: ProjectCard,
  argTypes: {
    project: { control: 'object' },
  },
} satisfies Meta;

const Template: StoryFn<typeof ProjectCard> = (args) => (
  <ProjectCard {...args} />
);

export const Default = Template.bind({});
Default.args = {
  project: {
    id: '5asdasadsa',
    createdAt: '2022-01-01',
    updatedAt: '2022-01-01',
    title: 'My Awesome Project',
    slug: 'my-awesome-project',
    shortDescription: 'A short description of the project.',
    description: {
      root: {
        direction: 'ltr',
        format: '',
        indent: 0,
        type: 'root',
        version: 1,
        children: [
          {
            detail: 0,
            format: 0,
            mode: 'normal',
            style: '',
            text: 'Studied advanced computer science topics, including algorithms, data structures, and machine learning.',
            type: 'text',
            version: 1,
          },
        ],
      },
    },
    technologies: [
      {
        id: '121aaaaa23',
        createdAt: '2022-01-01',
        updatedAt: '2022-01-01',
        name: 'React',
        icon: `<svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M847.7 112H176.3c-35.5 0-64.3 28.8-64.3 64.3v671.4c0 35.5 28.8 64.3 64.3 64.3h671.4c35.5 0 64.3-28.8 64.3-64.3V176.3c0-35.5-28.8-64.3-64.3-64.3zm0 736c-447.8-.1-671.7-.2-671.7-.3.1-447.8.2-671.7.3-671.7 447.8.1 671.7.2 671.7.3-.1 447.8-.2 671.7-.3 671.7zM230.6 411.9h118.7v381.8H230.6zm59.4-52.2c37.9 0 68.8-30.8 68.8-68.8a68.8 68.8 0 1 0-137.6 0c-.1 38 30.7 68.8 68.8 68.8zm252.3 245.1c0-49.8 9.5-98 71.2-98 60.8 0 61.7 56.9 61.7 101.2v185.7h118.6V584.3c0-102.8-22.2-181.9-142.3-181.9-57.7 0-96.4 31.7-112.3 61.7h-1.6v-52.2H423.7v381.8h118.6V604.8z"></path></svg>`,
      },
      {
        id: '12121213',
        createdAt: '2022-01-01',
        updatedAt: '2022-01-01',
        name: 'Next.js',
        icon: `<svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M847.7 112H176.3c-35.5 0-64.3 28.8-64.3 64.3v671.4c0 35.5 28.8 64.3 64.3 64.3h671.4c35.5 0 64.3-28.8 64.3-64.3V176.3c0-35.5-28.8-64.3-64.3-64.3zm0 736c-447.8-.1-671.7-.2-671.7-.3.1-447.8.2-671.7.3-671.7 447.8.1 671.7.2 671.7.3-.1 447.8-.2 671.7-.3 671.7zM230.6 411.9h118.7v381.8H230.6zm59.4-52.2c37.9 0 68.8-30.8 68.8-68.8a68.8 68.8 0 1 0-137.6 0c-.1 38 30.7 68.8 68.8 68.8zm252.3 245.1c0-49.8 9.5-98 71.2-98 60.8 0 61.7 56.9 61.7 101.2v185.7h118.6V584.3c0-102.8-22.2-181.9-142.3-181.9-57.7 0-96.4 31.7-112.3 61.7h-1.6v-52.2H423.7v381.8h118.6V604.8z"></path></svg>`,
      },
      {
        id: '12123',
        createdAt: '2022-01-01',
        updatedAt: '2022-01-01',
        name: 'TypeScript',
        icon: `<svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M847.7 112H176.3c-35.5 0-64.3 28.8-64.3 64.3v671.4c0 35.5 28.8 64.3 64.3 64.3h671.4c35.5 0 64.3-28.8 64.3-64.3V176.3c0-35.5-28.8-64.3-64.3-64.3zm0 736c-447.8-.1-671.7-.2-671.7-.3.1-447.8.2-671.7.3-671.7 447.8.1 671.7.2 671.7.3-.1 447.8-.2 671.7-.3 671.7zM230.6 411.9h118.7v381.8H230.6zm59.4-52.2c37.9 0 68.8-30.8 68.8-68.8a68.8 68.8 0 1 0-137.6 0c-.1 38 30.7 68.8 68.8 68.8zm252.3 245.1c0-49.8 9.5-98 71.2-98 60.8 0 61.7 56.9 61.7 101.2v185.7h118.6V584.3c0-102.8-22.2-181.9-142.3-181.9-57.7 0-96.4 31.7-112.3 61.7h-1.6v-52.2H423.7v381.8h118.6V604.8z"></path></svg>`,
      },
    ],
    thumbnail: {
      id: '5sdsa',
      createdAt: '2022-01-01',
      updatedAt: '2022-01-01',
      url: 'https://via.placeholder.com/420x304',
    },
  },
};

export const WithLiveAndGithubLinks = Template.bind({});
WithLiveAndGithubLinks.args = {
  project: {
    id: '5sadsa',
    createdAt: '2022-01-01',
    updatedAt: '2022-01-01',
    title: 'Another Great Project',
    slug: 'another-great-project',
    shortDescription: 'This project has live and GitHub links.',
    description: {
      root: {
        direction: 'ltr',
        format: '',
        indent: 0,
        type: 'root',
        version: 1,
        children: [
          {
            detail: 0,
            format: 0,
            mode: 'normal',
            style: '',
            text: 'Studied advanced computer science topics, including algorithms, data structures, and machine learning.',
            type: 'text',
            version: 1,
          },
        ],
      },
    },
    technologies: [
      {
        id: '1211212323',
        createdAt: '2022-01-01',
        updatedAt: '2022-01-01',
        name: 'Node.js',
        icon: `<svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M847.7 112H176.3c-35.5 0-64.3 28.8-64.3 64.3v671.4c0 35.5 28.8 64.3 64.3 64.3h671.4c35.5 0 64.3-28.8 64.3-64.3V176.3c0-35.5-28.8-64.3-64.3-64.3zm0 736c-447.8-.1-671.7-.2-671.7-.3.1-447.8.2-671.7.3-671.7 447.8.1 671.7.2 671.7.3-.1 447.8-.2 671.7-.3 671.7zM230.6 411.9h118.7v381.8H230.6zm59.4-52.2c37.9 0 68.8-30.8 68.8-68.8a68.8 68.8 0 1 0-137.6 0c-.1 38 30.7 68.8 68.8 68.8zm252.3 245.1c0-49.8 9.5-98 71.2-98 60.8 0 61.7 56.9 61.7 101.2v185.7h118.6V584.3c0-102.8-22.2-181.9-142.3-181.9-57.7 0-96.4 31.7-112.3 61.7h-1.6v-52.2H423.7v381.8h118.6V604.8z"></path></svg>`,
      },
      {
        id: '12',
        createdAt: '2022-01-01',
        updatedAt: '2022-01-01',
        name: 'GraphQL',
        icon: `<svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M847.7 112H176.3c-35.5 0-64.3 28.8-64.3 64.3v671.4c0 35.5 28.8 64.3 64.3 64.3h671.4c35.5 0 64.3-28.8 64.3-64.3V176.3c0-35.5-28.8-64.3-64.3-64.3zm0 736c-447.8-.1-671.7-.2-671.7-.3.1-447.8.2-671.7.3-671.7 447.8.1 671.7.2 671.7.3-.1 447.8-.2 671.7-.3 671.7zM230.6 411.9h118.7v381.8H230.6zm59.4-52.2c37.9 0 68.8-30.8 68.8-68.8a68.8 68.8 0 1 0-137.6 0c-.1 38 30.7 68.8 68.8 68.8zm252.3 245.1c0-49.8 9.5-98 71.2-98 60.8 0 61.7 56.9 61.7 101.2v185.7h118.6V584.3c0-102.8-22.2-181.9-142.3-181.9-57.7 0-96.4 31.7-112.3 61.7h-1.6v-52.2H423.7v381.8h118.6V604.8z"></path></svg>`,
      },
      {
        id: '13',
        createdAt: '2022-01-01',
        updatedAt: '2022-01-01',
        name: 'TailwindCSS',
        icon: `<svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M847.7 112H176.3c-35.5 0-64.3 28.8-64.3 64.3v671.4c0 35.5 28.8 64.3 64.3 64.3h671.4c35.5 0 64.3-28.8 64.3-64.3V176.3c0-35.5-28.8-64.3-64.3-64.3zm0 736c-447.8-.1-671.7-.2-671.7-.3.1-447.8.2-671.7.3-671.7 447.8.1 671.7.2 671.7.3-.1 447.8-.2 671.7-.3 671.7zM230.6 411.9h118.7v381.8H230.6zm59.4-52.2c37.9 0 68.8-30.8 68.8-68.8a68.8 68.8 0 1 0-137.6 0c-.1 38 30.7 68.8 68.8 68.8zm252.3 245.1c0-49.8 9.5-98 71.2-98 60.8 0 61.7 56.9 61.7 101.2v185.7h118.6V584.3c0-102.8-22.2-181.9-142.3-181.9-57.7 0-96.4 31.7-112.3 61.7h-1.6v-52.2H423.7v381.8h118.6V604.8z"></path></svg>`,
      },
    ],
    thumbnail: {
      url: 'https://via.placeholder.com/420x304',
      id: '12123',
      createdAt: '2022-01-01',
      updatedAt: '2022-01-01',
    },
    liveProjectUrl: 'https://live-project.example.com',
    githubUrl: 'https://github.com/example/repo',
  },
};
