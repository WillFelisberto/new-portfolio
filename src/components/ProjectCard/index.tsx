'use client';
import Image from 'next/image';
import { HiArrowNarrowRight } from 'react-icons/hi';
import { motion } from 'framer-motion';
import TechBadge from '../TechBadge';
import Link from '@/components/Link';
import { fadeUpAnimation, techBadgeAnimation } from '@/lib/animations';
import { DescriptionType } from '../RichText';
import { KnownTechProps } from '../KnownTech';

type ProjectCardProps = {
  project: Project;
};

export type Project = {
  title: string;
  slug: string;
  description: DescriptionType;
  shortDescription: string;
  technologies: KnownTechProps['tech'][];
  pageThumbnail?: {
    url: string;
  };
  thumbnail?: {
    url: string;
  };
  liveProjectUrl?: string;
  githubUrl?: string;
  sections?: {
    title: string;
    description: DescriptionType;
    image: {
      url: string;
      alt: string;
    };
  }[];
};

const ProjectCard = ({ project }: ProjectCardProps) => {
  return (
    <motion.div
      className="flex flex-col gap-6 lg:flex-row lg:gap-12"
      initial={{ opacity: 0, y: 100 }}
      whileInView={{ opacity: 1, y: 0 }}
      exit={{ opacity: 0, y: 100 }}
      transition={{ duration: 0.5 }}
    >
      <motion.div
        className="h-[200px] w-full sm:h-[300px] lg:min-h-full lg:w-[420px]"
        initial={{ opacity: 0, y: 100, scale: 0.5 }}
        whileInView={{ opacity: 1, y: 0, scale: 1 }}
        exit={{ opacity: 0, y: 100, scale: 0.5 }}
        transition={{ duration: 0.3, delay: 0.3 }}
      >
        <Link href={`/projects/${project.slug}`}>
          <Image
            src={project.thumbnail ? project.thumbnail.url : ''}
            width={420}
            height={304}
            alt={`Thumbnail do projeto ${project.title}`}
            className="h-full w-full rounded-lg object-cover"
          />
        </Link>
      </motion.div>

      <div className="flex-1 lg:py-[18px]">
        <Link href={`/projects/${project.slug}`}>
          <motion.h3
            className="flex items-center gap-3 text-lg font-medium text-gray-50"
            {...fadeUpAnimation}
            transition={{ duration: 0.7 }}
          >
            <Image
              width={20}
              height={20}
              alt="thumbnail project book"
              src="/images/icons/project-title-icon.svg"
            />
            {project.title}
          </motion.h3>
        </Link>
        <Link href={`/projects/${project.slug}`}>
          <motion.p
            className="my-6 text-gray-400"
            {...fadeUpAnimation}
            transition={{ duration: 0.2, delay: 0.3 }}
          >
            {project.shortDescription}
          </motion.p>
        </Link>

        <div className="mb-8 flex flex-wrap gap-x-2 gap-y-3 lg:max-w-[350px]">
          {project.technologies.map((tech, i) => (
            <TechBadge
              name={tech.name}
              key={`${project.title}-tech-${tech.name}`}
              {...techBadgeAnimation}
              transition={{ duration: 0.2, delay: 0.5 + i * 0.1 }}
            />
          ))}
        </div>

        <Link href={`/projects/${project.slug}`}>
          See project
          <HiArrowNarrowRight />
        </Link>
      </div>
    </motion.div>
  );
};

export default ProjectCard;
