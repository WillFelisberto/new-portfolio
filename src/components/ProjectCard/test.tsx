import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import ProjectCard, { Project } from '.';

const mockDefaultProject: Project = {
  title: 'My Awesome Project',
  slug: 'my-awesome-project',
  shortDescription: 'A short description of the project.',
  description: {
    root: {
      direction: 'ltr',
      format: '',
      indent: 0,
      type: 'root',
      version: 1,
      children: [
        {
          detail: 0,
          format: 0,
          mode: 'normal',
          style: '',
          text: 'Studied advanced computer science topics, including algorithms, data structures, and machine learning.',
          type: 'text',
          version: 1,
        },
      ],
    },
  },
  technologies: [
    { name: 'React', icon: '<svg></svg>' },
    { name: 'Next.js', icon: '<svg></svg>' },
    { name: 'TypeScript', icon: '<svg></svg>' },
  ],
  thumbnail: { url: 'https://via.placeholder.com/420x304' },
};

describe('ProjectCard Component', () => {
  it('renders the project title and description', () => {
    const { container } = render(<ProjectCard project={mockDefaultProject} />);

    const title = screen.getByRole('heading', { name: /my awesome project/i });
    expect(title).toBeInTheDocument();

    const description = screen.getByText(
      /a short description of the project\./i,
    );
    expect(description).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('renders the project thumbnail', () => {
    render(<ProjectCard project={mockDefaultProject} />);

    const thumbnail = screen.getByAltText(
      /thumbnail do projeto my awesome project/i,
    );
    expect(thumbnail).toBeInTheDocument();
  });

  it('renders technology badges', () => {
    render(<ProjectCard project={mockDefaultProject} />);

    mockDefaultProject.technologies.forEach((tech) => {
      const badge = screen.getByText(tech.name);
      expect(badge).toBeInTheDocument();
    });
  });
});
