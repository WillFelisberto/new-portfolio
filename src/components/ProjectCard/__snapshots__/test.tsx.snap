// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCard Component renders the project title and description 1`] = `
<div>
  <div
    class="flex flex-col gap-6 lg:flex-row lg:gap-12"
    style="opacity: 0; transform: translateY(100px);"
  >
    <div
      class="h-[200px] w-full sm:h-[300px] lg:min-h-full lg:w-[420px]"
      style="opacity: 0; transform: translateY(100px) scale(0.5);"
    >
      <a
        class="flex items-center gap-2 text-sm text-gray-300 transition-colors hover:text-sky-500"
        href="/projects/my-awesome-project"
      >
        <img
          alt="Thumbnail do projeto My Awesome Project"
          class="h-full w-full rounded-lg object-cover"
          data-nimg="1"
          decoding="async"
          height="304"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fvia.placeholder.com%2F420x304&w=1080&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fvia.placeholder.com%2F420x304&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fvia.placeholder.com%2F420x304&w=1080&q=75 2x"
          style="color: transparent;"
          width="420"
        />
      </a>
    </div>
    <div
      class="flex-1 lg:py-[18px]"
    >
      <a
        class="flex items-center gap-2 text-sm text-gray-300 transition-colors hover:text-sky-500"
        href="/projects/my-awesome-project"
      >
        <h3
          class="flex items-center gap-3 text-lg font-medium text-gray-50"
          style="opacity: 0; transform: translateY(50px);"
        >
          <img
            alt="thumbnail project book"
            data-nimg="1"
            decoding="async"
            height="20"
            loading="lazy"
            src="/images/icons/project-title-icon.svg"
            style="color: transparent;"
            width="20"
          />
          My Awesome Project
        </h3>
      </a>
      <a
        class="flex items-center gap-2 text-sm text-gray-300 transition-colors hover:text-sky-500"
        href="/projects/my-awesome-project"
      >
        <p
          class="my-6 text-gray-400"
          style="opacity: 0; transform: translateY(50px);"
        >
          A short description of the project.
        </p>
      </a>
      <div
        class="mb-8 flex flex-wrap gap-x-2 gap-y-3 lg:max-w-[350px]"
      >
        <span
          class="rounded-lg bg-sky-900/80 px-3 py-1 text-sm text-sky-400"
          style="opacity: 0; transform: scale(0);"
        >
          React
        </span>
        <span
          class="rounded-lg bg-sky-900/80 px-3 py-1 text-sm text-sky-400"
          style="opacity: 0; transform: scale(0);"
        >
          Next.js
        </span>
        <span
          class="rounded-lg bg-sky-900/80 px-3 py-1 text-sm text-sky-400"
          style="opacity: 0; transform: scale(0);"
        >
          TypeScript
        </span>
      </div>
      <a
        class="flex items-center gap-2 text-sm text-gray-300 transition-colors hover:text-sky-500"
        href="/projects/my-awesome-project"
      >
        See project
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.293 5.293a1 1 0 011.414 0l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414-1.414L14.586 11H3a1 1 0 110-2h11.586l-2.293-2.293a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;
